File: packages/cortex-dart/test/errors_test.dart
import 'package:nfinitmonkeys_cortex_sdk/cortex_sdk.dart';
import 'package:test/test.dart';

void main() {
  group('CortexException', () {
    test('has message, code, and statusCode', () {
      const e = CortexException(
        message: 'Something went wrong',
        code: 'ERR_01',
        statusCode: 400,
      );
      expect(e.message, 'Something went wrong');
      expect(e.code, 'ERR_01');
      expect(e.statusCode, 400);
    });

    test('toString does not expose sensitive info', () {
      const e = CortexException(message: 'Bad request');
      expect(e.toString(), 'CortexException(Bad request)');
    });

    test('equality', () {
      const a = CortexException(message: 'error', statusCode: 400);
      const b = CortexException(message: 'error', statusCode: 400);
      const c = CortexException(message: 'different', statusCode: 400);
      expect(a, equals(b));
      expect(a, isNot(equals(c)));
      expect(a.hashCode, b.hashCode);
    });
  });

  group('CortexAuthenticationException', () {
    test('defaults', () {
      const e = CortexAuthenticationException();
      expect(e.statusCode, 401);
      expect(e.message, contains('Authentication'));
      expect(e.toString(), contains('CortexAuthenticationException'));
    });

    test('custom message', () {
      const e = CortexAuthenticationException(message: 'Invalid token');
      expect(e.message, 'Invalid token');
    });
  });

  group('CortexForbiddenException', () {
    test('defaults', () {
      const e = CortexForbiddenException();
      expect(e.statusCode, 403);
    });
  });

  group('CortexNotFoundException', () {
    test('defaults', () {
      const e = CortexNotFoundException();
      expect(e.statusCode, 404);
    });
  });

  group('CortexRateLimitException', () {
    test('includes retry-after', () {
      const e = CortexRateLimitException(
        retryAfter: Duration(seconds: 30),
      );
      expect(e.statusCode, 429);
      expect(e.retryAfter, const Duration(seconds: 30));
    });

    test('equality includes retryAfter', () {
      const a = CortexRateLimitException(retryAfter: Duration(seconds: 30));
      const b = CortexRateLimitException(retryAfter: Duration(seconds: 30));
      const c = CortexRateLimitException(retryAfter: Duration(seconds: 60));
      expect(a, equals(b));
      expect(a, isNot(equals(c)));
    });
  });

  group('CortexServerException', () {
    test('defaults to 500', () {
      const e = CortexServerException();
      expect(e.statusCode, 500);
    });

    test('custom status code', () {
      const e = CortexServerException(statusCode: 503);
      expect(e.statusCode, 503);
    });
  });

  group('CortexValidationException', () {
    test('includes parameter name', () {
      const e = CortexValidationException(
        message: 'Must not be empty',
        parameter: 'model',
      );
      expect(e.parameter, 'model');
      expect(e.toString(), contains('model'));
      expect(e.toString(), contains('Must not be empty'));
    });

    test('equality includes parameter', () {
      const a = CortexValidationException(
        message: 'err',
        parameter: 'model',
      );
      const b = CortexValidationException(
        message: 'err',
        parameter: 'model',
      );
      const c = CortexValidationException(
        message: 'err',
        parameter: 'prompt',
      );
      expect(a, equals(b));
      expect(a, isNot(equals(c)));
    });
  });

  group('CortexTimeoutException', () {
    test('defaults', () {
      const e = CortexTimeoutException();
      expect(e.message, contains('timed out'));
    });
  });

  group('CortexConnectionException', () {
    test('defaults', () {
      const e = CortexConnectionException();
      expect(e.message, contains('Connection'));
    });
  });

  group('CortexStreamException', () {
    test('defaults', () {
      const e = CortexStreamException();
      expect(e.message, contains('Stream'));
    });
  });

  group('exceptionFromStatusCode', () {
    test('maps 401 to CortexAuthenticationException', () {
      final e = exceptionFromStatusCode(401, 'Unauthorized');
      expect(e, isA<CortexAuthenticationException>());
      expect(e.message, 'Unauthorized');
    });

    test('maps 403 to CortexForbiddenException', () {
      final e = exceptionFromStatusCode(403, 'Forbidden');
      expect(e, isA<CortexForbiddenException>());
    });

    test('maps 404 to CortexNotFoundException', () {
      final e = exceptionFromStatusCode(404, 'Not found');
      expect(e, isA<CortexNotFoundException>());
    });

    test('maps 429 to CortexRateLimitException with retryAfter', () {
      final e = exceptionFromStatusCode(
        429,
        'Rate limited',
        retryAfter: const Duration(seconds: 10),
      );
      expect(e, isA<CortexRateLimitException>());
      expect((e as CortexRateLimitException).retryAfter,
          const Duration(seconds: 10));
    });

    test('maps 500 to CortexServerException', () {
      final e = exceptionFromStatusCode(500, 'Server error');
      expect(e, isA<CortexServerException>());
    });

    test('maps 502 to CortexServerException', () {
      final e = exceptionFromStatusCode(502, 'Bad gateway');
      expect(e, isA<CortexServerException>());
      expect(e.statusCode, 502);
    });

    test('maps unknown status to CortexException', () {
      final e = exceptionFromStatusCode(422, 'Unprocessable');
      expect(e, isA<CortexException>());
      expect(e, isNot(isA<CortexServerException>()));
      expect(e.statusCode, 422);
    });

    test('uses HTTP status as message when body is empty', () {
      final e = exceptionFromStatusCode(500, '');
      expect(e.message, 'HTTP 500');
    });
  });
}

File: packages/cortex-js/tests/errors.test.ts
import { describe, it, expect } from 'vitest';
import {
  CortexError,
  AuthenticationError,
  PermissionDeniedError,
  NotFoundError,
  RateLimitError,
  ValidationError,
  TimeoutError,
  ConnectionError,
  ServerError,
  errorFromStatus,
} from '../src/errors.js';

describe('Errors', () => {
  describe('CortexError', () => {
    it('creates an error with message', () => {
      const error = new CortexError('Something went wrong');
      expect(error.message).toBe('Something went wrong');
      expect(error.name).toBe('CortexError');
      expect(error).toBeInstanceOf(Error);
    });

    it('includes status and code', () => {
      const error = new CortexError('Bad request', { status: 400, code: 'bad_request' });
      expect(error.status).toBe(400);
      expect(error.code).toBe('bad_request');
    });

    it('includes requestId', () => {
      const error = new CortexError('Error', { requestId: 'req-123' });
      expect(error.requestId).toBe('req-123');
    });

    it('sanitizes API keys from messages', () => {
      const error = new CortexError(
        'Failed with key sk-cortex-abc123def456 in request',
      );
      expect(error.message).not.toContain('sk-cortex-abc123def456');
      expect(error.message).toContain('sk-cortex-***REDACTED***');
    });

    it('sanitizes multiple API keys', () => {
      const error = new CortexError(
        'Keys: sk-cortex-first and sk-cortex-second',
      );
      expect(error.message).not.toContain('sk-cortex-first');
      expect(error.message).not.toContain('sk-cortex-second');
      expect(error.message).toMatch(/sk-cortex-\*\*\*REDACTED\*\*\*.*sk-cortex-\*\*\*REDACTED\*\*\*/);
    });
  });

  describe('AuthenticationError', () => {
    it('has correct defaults', () => {
      const error = new AuthenticationError();
      expect(error.message).toBe('Invalid or missing API key');
      expect(error.status).toBe(401);
      expect(error.code).toBe('authentication_error');
      expect(error.name).toBe('AuthenticationError');
    });

    it('is instanceof CortexError', () => {
      expect(new AuthenticationError()).toBeInstanceOf(CortexError);
    });
  });

  describe('PermissionDeniedError', () => {
    it('has correct defaults', () => {
      const error = new PermissionDeniedError();
      expect(error.status).toBe(403);
      expect(error.name).toBe('PermissionDeniedError');
    });
  });

  describe('NotFoundError', () => {
    it('has correct defaults', () => {
      const error = new NotFoundError();
      expect(error.status).toBe(404);
      expect(error.name).toBe('NotFoundError');
    });
  });

  describe('RateLimitError', () => {
    it('has correct defaults', () => {
      const error = new RateLimitError();
      expect(error.status).toBe(429);
      expect(error.code).toBe('rate_limit_exceeded');
      expect(error.name).toBe('RateLimitError');
    });

    it('includes retryAfter', () => {
      const error = new RateLimitError('Too many requests', 30);
      expect(error.retryAfter).toBe(30);
    });
  });

  describe('ValidationError', () => {
    it('includes field name', () => {
      const error = new ValidationError('Invalid model', 'model');
      expect(error.field).toBe('model');
      expect(error.code).toBe('validation_error');
      expect(error.name).toBe('ValidationError');
    });
  });

  describe('TimeoutError', () => {
    it('has correct defaults', () => {
      const error = new TimeoutError();
      expect(error.message).toBe('Request timed out');
      expect(error.code).toBe('timeout');
      expect(error.name).toBe('TimeoutError');
    });
  });

  describe('ConnectionError', () => {
    it('preserves cause', () => {
      const cause = new TypeError('fetch failed');
      const error = new ConnectionError('Connection failed', cause);
      expect(error.cause).toBe(cause);
      expect(error.name).toBe('ConnectionError');
    });
  });

  describe('ServerError', () => {
    it('has correct status', () => {
      const error = new ServerError('Internal server error', 502, 'req-456');
      expect(error.status).toBe(502);
      expect(error.requestId).toBe('req-456');
      expect(error.name).toBe('ServerError');
    });
  });

  describe('errorFromStatus', () => {
    it('maps 401 to AuthenticationError', () => {
      expect(errorFromStatus(401, 'Unauthorized')).toBeInstanceOf(AuthenticationError);
    });

    it('maps 403 to PermissionDeniedError', () => {
      expect(errorFromStatus(403, 'Forbidden')).toBeInstanceOf(PermissionDeniedError);
    });

    it('maps 404 to NotFoundError', () => {
      expect(errorFromStatus(404, 'Not found')).toBeInstanceOf(NotFoundError);
    });

    it('maps 429 to RateLimitError', () => {
      const headers = new Headers({ 'retry-after': '10' });
      const error = errorFromStatus(429, 'Rate limited', headers);
      expect(error).toBeInstanceOf(RateLimitError);
      expect((error as RateLimitError).retryAfter).toBe(10);
    });

    it('maps 500+ to ServerError', () => {
      expect(errorFromStatus(500, 'Internal')).toBeInstanceOf(ServerError);
      expect(errorFromStatus(502, 'Bad Gateway')).toBeInstanceOf(ServerError);
      expect(errorFromStatus(503, 'Unavailable')).toBeInstanceOf(ServerError);
    });

    it('maps unknown status to CortexError', () => {
      const error = errorFromStatus(418, 'Teapot');
      expect(error).toBeInstanceOf(CortexError);
      expect(error.status).toBe(418);
    });
  });
});
